Tasklist importer cannot deserialize user tasks variables with Long values

### Describe the bug

<!-- A clear and concise description of what the problem/bug is about and what is the current behavior. -->
Reported in this [Slack discussion](https://camunda.slack.com/archives/CSQ2E3BT4/p1741049329356789)
**Update:** also reported in this support ticket https://jira.camunda.com/browse/SUPPORT-26412

**Affects Tasklist 8.6 and 8.7**

### To Reproduce

Steps to reproduce the behavior:
1. Start a process instance with Camunda user task
2. Complete the Camunda user task with Zeebe API using a Long variable
```
curl --location 'http://localhost:8089/v2/user-tasks/4503599627370503/completion' \
--header 'Content-Type: application/json' \
--data '{
    "variables": {
        "jobKey": 4503599627370503
    }
}'
```

importer fails to import the Completed user task record with the following error:

```
2025-03-12 14:26:36.262 [] [import_2] [] ERROR
      io.camunda.tasklist.zeebeimport.ImportJob - Failed to parse variable jobKey=4503599627370503
io.camunda.tasklist.exceptions.TasklistRuntimeException: Failed to parse variable jobKey=4503599627370503
        at io.camunda.tasklist.zeebeimport.v860.processors.common.UserTaskRecordToVariableEntityMapper.mapVariables(UserTaskRecordToVariableEntityMapper.java:48) ~[tasklist-importer-860-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.v860.processors.es.UserTaskZeebeRecordProcessorElasticSearch.processUserTaskRecord(UserTaskZeebeRecordProcessorElasticSearch.java:67) ~[tasklist-importer-860-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.v860.processors.es.BulkProcessorElasticSearch.processZeebeRecords(BulkProcessorElasticSearch.java:103) ~[tasklist-importer-860-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.es.AbstractImportBatchProcessorElasticSearch.performImport(AbstractImportBatchProcessorElasticSearch.java:32) ~[tasklist-importer-common-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.ImportJobAbstract.processOneIndexBatch(ImportJobAbstract.java:89) ~[tasklist-importer-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.ImportJobAbstract.call(ImportJobAbstract.java:59) ~[tasklist-importer-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.ImportJobAbstract.call(ImportJobAbstract.java:24) ~[tasklist-importer-8.6.10.jar:8.6.10]
        at io.camunda.tasklist.zeebeimport.RecordsReaderAbstract.lambda$executeJob$1(RecordsReaderAbstract.java:225) ~[tasklist-importer-8.6.10.jar:8.6.10]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Cannot invoke "com.fasterxml.jackson.core.JsonStreamContext.getCurrentValue()" because the return value of "com.fasterxml.jackson.core.JsonStreamContext.getParent()" is null
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._wrapAsIOE(DefaultSerializerProvider.java:531) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:504) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:341) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4811) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4052) ~[jackson-databind-2.18.2.jar:2.18.2]
        at io.camunda.tasklist.zeebeimport.v860.processors.common.UserTaskRecordToVariableEntityMapper.mapVariables(UserTaskRecordToVariableEntityMapper.java:46) ~[tasklist-importer-860-8.6.10.jar:8.6.10]
        ... 11 more
Caused by: java.lang.NullPointerException: Cannot invoke "com.fasterxml.jackson.core.JsonStreamContext.getCurrentValue()" because the return value of "com.fasterxml.jackson.core.JsonStreamContext.getParent()" is null
        at io.camunda.zeebe.gateway.rest.serializer.LongSerializer.gatewayRestPackage(LongSerializer.java:44) ~[zeebe-gateway-rest-8.6.10.jar:8.6.10]
        at io.camunda.zeebe.gateway.rest.serializer.LongSerializer.serialize(LongSerializer.java:59) ~[zeebe-gateway-rest-8.6.10.jar:8.6.10]
        at io.camunda.zeebe.gateway.rest.serializer.LongSerializer.serialize(LongSerializer.java:20) ~[zeebe-gateway-rest-8.6.10.jar:8.6.10]
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:341) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4811) ~[jackson-databind-2.18.2.jar:2.18.2]
        at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4052) ~[jackson-databind-2.18.2.jar:2.18.2]
        at io.camunda.tasklist.zeebeimport.v860.processors.common.UserTaskRecordToVariableEntityMapper.mapVariables(UserTaskRecordToVariableEntityMapper.java:46) ~[tasklist-importer-860-8.6.10.jar:8.6.10]
```

### Root cause
- [UserTaskRecordToVariableEntityMapper](https://github.com/camunda/camunda/blob/0a9b7a03f0b20b74e238d95f0f240422e92057ac/tasklist/importer-860/src/main/java/io/camunda/tasklist/zeebeimport/v860/processors/common/UserTaskRecordToVariableEntityMapper.java#L34) (in Tasklist importer) uses the [default object mapper](https://github.com/camunda/camunda/blob/0a9b7a03f0b20b74e238d95f0f240422e92057ac/dist/src/main/java/io/camunda/application/commons/sources/DefaultObjectMapperConfiguration.java#L47)  instead of the tasklistObjectMapper, as it does not qualify the autowired objectMapper
- The default object mapper uses [`gatewayRestObjectMapperCustomizer`](https://github.com/camunda/camunda/blob/0a9b7a03f0b20b74e238d95f0f240422e92057ac/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/config/JacksonConfig.java#L22) which defines the [LongSerializer](https://github.com/camunda/camunda/blob/stable/8.6/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/serializer/LongSerializer.java) that throws the NPE shown in the stacktrace
- Adding a qualifier on the objectMapper used by the importer should fix the issue
- We need to understand what is the impact of this NPE in [`LongSerializer`](https://github.com/camunda/camunda/blob/stable/8.6/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/serializer/LongSerializer.java). Add a null check as discussed in [Slack](https://camunda.slack.com/archives/C06UKS51QV9/p1741795221605539)

--- a/tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/common/UserTaskRecordToVariableEntityMapper.java
+++ b/tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/processors/common/UserTaskRecordToVariableEntityMapper.java
@@ -21,6 +21,7 @@ import java.util.Map;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Component;
 
 @Component
@@ -31,7 +32,9 @@ public class UserTaskRecordToVariableEntityMapper {
 
   @Autowired private TasklistProperties tasklistProperties;
 
-  @Autowired private ObjectMapper objectMapper;
+  @Autowired
+  @Qualifier("tasklistObjectMapper")
+  private ObjectMapper objectMapper;
 
   public List<TaskVariableEntity> mapVariables(final Record<UserTaskRecordValue> record) {
     final List<TaskVariableEntity> variables = new ArrayList<>();

--- a/tasklist/importer-860/src/main/java/io/camunda/tasklist/zeebeimport/v860/processors/common/UserTaskRecordToVariableEntityMapper.java
+++ b/tasklist/importer-860/src/main/java/io/camunda/tasklist/zeebeimport/v860/processors/common/UserTaskRecordToVariableEntityMapper.java
@@ -21,6 +21,7 @@ import java.util.Map;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Component;
 
 @Component
@@ -31,7 +32,9 @@ public class UserTaskRecordToVariableEntityMapper {
 
   @Autowired private TasklistProperties tasklistProperties;
 
-  @Autowired private ObjectMapper objectMapper;
+  @Autowired
+  @Qualifier("tasklistObjectMapper")
+  private ObjectMapper objectMapper;
 
   public List<TaskVariableEntity> mapVariables(final Record<UserTaskRecordValue> record) {
     final List<TaskVariableEntity> variables = new ArrayList<>();

--- a/tasklist/qa/integration-tests/pom.xml
+++ b/tasklist/qa/integration-tests/pom.xml
@@ -490,6 +490,24 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>com.tngtech.archunit</groupId>
+      <artifactId>archunit</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>com.tngtech.archunit</groupId>
+      <artifactId>archunit-junit5-api</artifactId>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>com.tngtech.archunit</groupId>
+      <artifactId>archunit-junit5-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+
   </dependencies>
 
   <build>

--- a/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/ObjectMapperQualifierArchTest.java
+++ b/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/ObjectMapperQualifierArchTest.java
@@ -1,0 +1,65 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.tasklist;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.fields;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.tngtech.archunit.core.domain.JavaField;
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+@AnalyzeClasses(
+    packages = "io.camunda.tasklist",
+    importOptions = ImportOption.DoNotIncludeTests.class)
+public class ObjectMapperQualifierArchTest {
+
+  @ArchTest
+  public static final ArchRule AUTOWIRED_OBJECT_MAPPER_FIELDS_SHOULD_HAVE_QUALIFIER =
+      fields()
+          .that()
+          .areAnnotatedWith(Autowired.class)
+          .and()
+          .haveRawType(ObjectMapper.class)
+          .should(
+              new ArchCondition<>("have @Qualifier(\"tasklistObjectMapper\")") {
+                @Override
+                public void check(final JavaField field, final ConditionEvents events) {
+                  final boolean hasQualifier = field.isAnnotatedWith(Qualifier.class);
+                  if (!hasQualifier) {
+                    events.add(
+                        SimpleConditionEvent.violated(
+                            field,
+                            "Field "
+                                + field.getFullName()
+                                + "in class "
+                                + field.getOwner().getFullName()
+                                + " is missing @Qualifier(\"tasklistObjectMapper\")"));
+                  } else if (!"tasklistObjectMapper"
+                      .equals(field.getAnnotationOfType(Qualifier.class).value())) {
+                    events.add(
+                        SimpleConditionEvent.violated(
+                            field,
+                            "Field "
+                                + field.getFullName()
+                                + "in class "
+                                + field.getOwner().getFullName()
+                                + " has @Qualifier(\"%s\")"
+                                    .formatted(
+                                        field.getAnnotationOfType(Qualifier.class).value())));
+                  }
+                }
+              });
+}

--- a/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/util/TestApplication.java
+++ b/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/util/TestApplication.java
@@ -7,6 +7,7 @@
  */
 package io.camunda.tasklist.util;
 
+import io.camunda.application.commons.CommonsModuleConfiguration;
 import io.camunda.tasklist.TasklistModuleConfiguration;
 import io.camunda.tasklist.data.DataGenerator;
 import io.camunda.tasklist.data.es.DevDataGeneratorElasticSearch;
@@ -34,7 +35,7 @@ import org.springframework.context.annotation.Profile;
           value = TasklistModuleConfiguration.class),
     },
     nameGenerator = FullyQualifiedAnnotationBeanNameGenerator.class)
-@Import({WebappsModuleConfiguration.class})
+@Import({WebappsModuleConfiguration.class, CommonsModuleConfiguration.class})
 public class TestApplication {
 
   public static void main(final String[] args) throws Exception {

--- a/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/zeebeimport/ZeebeUserTaskImportIT.java
+++ b/tasklist/qa/integration-tests/src/test/java/io/camunda/tasklist/zeebeimport/ZeebeUserTaskImportIT.java
@@ -83,7 +83,7 @@ public class ZeebeUserTaskImportIT extends TasklistZeebeIntegrationTest {
   }
 
   @Test
-  public void shouldImportZeebeUserTaskWithNullVariableValue() {
+  public void shouldImportCompletedZeebeUserTaskWithVariables() {
     final String bpmnProcessId = "testProcess";
     final String flowNodeBpmnId1 = "taskA";
     final String flowNodeBpmnId2 = "taskB";
@@ -106,7 +106,17 @@ public class ZeebeUserTaskImportIT extends TasklistZeebeIntegrationTest {
             .startProcessInstance(bpmnProcessId)
             .waitUntil()
             .taskIsCreated(flowNodeBpmnId1)
-            .completeZeebeUserTask(flowNodeBpmnId1, Map.of("varA", objectMapper.nullNode()))
+            .completeZeebeUserTask(
+                flowNodeBpmnId1,
+                Map.of(
+                    "varA",
+                    objectMapper.nullNode(),
+                    "varB",
+                    Long.MAX_VALUE,
+                    "varC",
+                    Integer.MAX_VALUE,
+                    "varD",
+                    "str"))
             .waitUntil()
             .taskIsCreated(flowNodeBpmnId2)
             .getTaskId();
@@ -123,7 +133,11 @@ public class ZeebeUserTaskImportIT extends TasklistZeebeIntegrationTest {
         .hasApplicationJsonContentType()
         .extractingListContent(objectMapper, VariableSearchResponse.class)
         .extracting("name", "previewValue", "value")
-        .containsExactly(tuple("varA", "null", "null"));
+        .containsExactlyInAnyOrder(
+            tuple("varA", "null", "null"),
+            tuple("varB", "9223372036854775807", "9223372036854775807"),
+            tuple("varC", "2147483647", "2147483647"),
+            tuple("varD", "\"str\"", "\"str\""));
   }
 
   @Test

--- a/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/serializer/LongSerializer.java
+++ b/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/serializer/LongSerializer.java
@@ -41,7 +41,8 @@ public class LongSerializer extends JsonSerializer<Long> {
 
   private boolean gatewayRestPackage(final JsonGenerator gen) {
     return gatewayRestPackage(gen.getOutputContext().getCurrentValue())
-        || gatewayRestPackage(gen.getOutputContext().getParent().getCurrentValue());
+        || (gen.getOutputContext().getParent() != null
+            && gatewayRestPackage(gen.getOutputContext().getParent().getCurrentValue()));
   }
 
   private boolean fieldNameEndsWithKey(final JsonGenerator gen) {

